crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::exit_codes::{CInputError, ExitCode};
+use num_complex::Complex64;
+use qiskit_accelerate::sparse_observable::{BitTerm, SparseObservable, SparseTermView};
+
+/// @ingroup QkSparseTerm
+/// A term in a [SparseObservable].
+///
+/// This term does not own the data, but only contains pointers (and a size) to the
+/// bit terms, indices and the coefficient.
+#[repr(C)]
+pub struct CSparseTerm {
+    coeff: *mut Complex64,
+    len: usize,
+    bit_terms: *mut BitTerm,
+    indices: *mut u32,
+    num_qubits: u32,
+}
+
+impl TryFrom<&CSparseTerm> for SparseTermView<'_> {
+    type Error = CInputError;
+
+    fn try_from(value: &CSparseTerm) -> Result<Self, Self::Error> {
+        if value.bit_terms.is_null() || value.indices.is_null() {
+            return Err(CInputError::NullPointerError);
+        }
+
+        if !value.bit_terms.is_aligned() || !value.indices.is_aligned() {
+            return Err(CInputError::AlignmentError);
+        }
+
+        let bit_terms = unsafe { ::std::slice::from_raw_parts(value.bit_terms, value.len) };
+        let indices = unsafe { ::std::slice::from_raw_parts(value.indices, value.len) };
+
+        Ok(SparseTermView {
+            num_qubits: value.num_qubits,
+            coeff: unsafe { *value.coeff },
+            bit_terms,
+            indices,
+        })
+    }
+}
+
+/// Check the pointer is not null and is aligned.
+fn check_ptr<T>(ptr: *const T) {
+    if ptr.is_null() {
+        panic!("Unexpected null pointer.");
+    };
+    if !ptr.is_aligned() {
+        panic!("Pointer is not properly aligned.");
+    };
+}
```

Comment:
Perhaps this should return `Result`, then we can use it in more places - when panicking is appropriate, we can `Result.unwrap` it to panic, and when we're using it fallibly (like above in the `SparseTerm` wrapper), we can match on the error result.

Reply:
Changed in https://github.com/Qiskit/qiskit/commit/bc376c6f27b20c932c7eb4e3da7e63ffb9cbf2ea